plugins/vpp/l3plugin/descriptor/vrrp.go

```diff
@@ -0,0 +1,235 @@
+		Retrieve:             ctx.Retrieve,
+		Dependencies:         ctx.Dependencies,
+		RetrieveDependencies: []string{ifdescriptor.InterfaceDescriptorName},
+	}
+
+	return adapter.NewVRRPEntryDescriptor(typedDescr)
+}
+
+// Validate returns error if given VRRP is not valid.
+func (d *VrrpDescriptor) Validate(key string, vrrp *l3.VRRPEntry) error {
+	if vrrp.Interface == "" {
+		return kvs.NewInvalidValueError(ErrMissingInterface, "interface")
+	}
+
+	if len(vrrp.Addrs) > maxUint8 || len(vrrp.Addrs) == 0 {
+		return kvs.NewInvalidValueError(ErrInvalidAddrNum, "addrs")
+	}
+
+	if vrrp.GetVrId() > maxUint8 || vrrp.GetVrId() == 0 {
+		return kvs.NewInvalidValueError(ErrInvalidVrID, "vr_id")
+	}
+
+	if vrrp.GetPriority() > maxUint8 || vrrp.GetPriority() == 0 {
+		return kvs.NewInvalidValueError(ErrInvalidPriority, "priority")
+	}
+
+	if vrrp.GetInterval() > maxUint16 || vrrp.GetInterval() == 0 {
+		return kvs.NewInvalidValueError(ErrInvalidInterval, "interval")
+	}
+
+	var ip net.IP
+	for _, addr := range vrrp.Addrs {
+		ip = net.ParseIP(addr)
+		if ip == nil {
+			return kvs.NewInvalidValueError(ErrInvalidVrrpIP, "addr")
+		}
+
+		if ip.To4() == nil && !vrrp.Ipv6 || ip.To4() != nil && vrrp.Ipv6 {
+			return kvs.NewInvalidValueError(ErrInvalidIPVer, "addr")
+		}
+	}
+	return nil
+}
+
+// Create adds VPP VRRP entry.
+func (d *VrrpDescriptor) Create(key string, vrrp *l3.VRRPEntry) (interface{}, error) {
+	if err := d.vrrpHandler.VppAddVrrp(vrrp); err != nil {
+		if errors.Is(vppcalls.ErrVRRPUnsupported, err) {
+			d.log.Debugf("Unsupported action: ", err)
+		}
+		return nil, err
+	}
+
+	if vrrp.Enabled {
+		if err := d.vrrpHandler.VppStartVrrp(vrrp); err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+// Delete removes VPP VRRP entry.
+func (d *VrrpDescriptor) Delete(key string, vrrp *l3.VRRPEntry, metadata interface{}) error {
+	if vrrp.Enabled {
+		if err := d.vrrpHandler.VppStopVrrp(vrrp); err != nil {
+			if errors.Is(vppcalls.ErrVRRPUnsupported, err) {
+				d.log.Debugf("Unsupported action: ", err)
+			}
+			return err
+		}
+	}
+	if err := d.vrrpHandler.VppDelVrrp(vrrp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// UpdateWithRecreate returns true if a VRRP update needs to be performed via re-crate.
+func (d *VrrpDescriptor) UpdateWithRecreate(_ string, oldVRRPEntry, newVRRPEntry *l3.VRRPEntry, _ interface{}) bool {
+	if oldVRRPEntry.Enabled == newVRRPEntry.Enabled {
+		// Something has changed, but it is not about enabled/disabled.
+		return true
+	}
+
+	if oldVRRPEntry.Interface == newVRRPEntry.Interface &&
+		oldVRRPEntry.Interval == newVRRPEntry.Interval &&
+		oldVRRPEntry.Priority == newVRRPEntry.Priority &&
+		oldVRRPEntry.VrId == newVRRPEntry.VrId &&
+		oldVRRPEntry.Ipv6 == newVRRPEntry.Ipv6 &&
+		oldVRRPEntry.Accept == newVRRPEntry.Accept &&
+		oldVRRPEntry.Preempt == newVRRPEntry.Preempt &&
+		oldVRRPEntry.Unicast == newVRRPEntry.Unicast &&
+		len(oldVRRPEntry.Addrs) == len(newVRRPEntry.Addrs) {
```

Comment:
What if some addresses changes, but overall count stays same? This `if` will return false before the next check that compares the addresses..